features/cellular/framework/mux/cellular_mux.cpp

```diff
@@ -0,0 +1,1237 @@
+            MBED_ASSERT(false);
+
+            break;
+    }
+
+    _state.is_dlci_open_running = 0;
+
+    return MUX_STATUS_SUCCESS;
+}
+
+
+void Mux::tx_retransmit_enqueu_entry_run()
+{
+    write_do();
+}
+
+
+Mux::MuxReturnStatus Mux::mux_start(Mux::MuxEstablishStatus &status)
+{
+    _mutex.lock();
+
+    if (_state.is_mux_open) {
+        _mutex.unlock();
+
+        return MUX_STATUS_NO_RESOURCE;
+    }
+    if (_state.is_mux_open_pending) {
+        _mutex.unlock();
+
+        return MUX_STATUS_INPROGRESS;
+    }
+    if (_state.is_mux_open_running ) {
+        _mutex.unlock();
+
+        return MUX_STATUS_INPROGRESS;
+    }
+
+    switch (_tx_context.tx_state) {
+            int32_t ret_wait;
+        case TX_IDLE:
+            /* Construct the frame, start the tx sequence, and suspend the call thread upon write sequence success. */
+            sabm_request_construct(0);
+            _tx_context.retransmit_counter = RETRANSMIT_COUNT;
+            tx_state_change(TX_RETRANSMIT_ENQUEUE, tx_retransmit_enqueu_entry_run, tx_idle_exit_run);
+            _state.is_mux_open_running = 1u;
+
+            _mutex.unlock();
+            ret_wait = _semaphore.wait();
+            MBED_ASSERT(ret_wait == 1);
+            status = static_cast<MuxEstablishStatus>(_shared_memory);
+            if (status == MUX_ESTABLISH_SUCCESS) {
+                _state.is_mux_open = 1u;
+            }
+
+            break;
+        case TX_INTERNAL_RESP:
+            _state.is_mux_open_pending = 1u;
+
+            _mutex.unlock();
+            ret_wait = _semaphore.wait();
+            MBED_ASSERT(ret_wait == 1);
+            status = static_cast<MuxEstablishStatus>(_shared_memory);
+            if (status == MUX_ESTABLISH_SUCCESS) {
+                _state.is_mux_open = 1u;
+            }
+
+            break;
+        default:
+            /* Code that should never be reached. */
+            MBED_ASSERT(false);
+            break;
+    };
+
+    _state.is_mux_open_running = 0;
+
+    return MUX_STATUS_SUCCESS;
+}
+
+
+void Mux::user_information_construct(uint8_t dlci_id, const void* buffer, size_t size)
+{
+    frame_hdr_t *frame_hdr =
+        reinterpret_cast<frame_hdr_t *>(&(Mux::_tx_context.buffer[FRAME_FLAG_SEQUENCE_FIELD_INDEX]));
+
+    frame_hdr->address  = EA_BIT | CR_BIT | (dlci_id << 2);
+    frame_hdr->control  = FRAME_TYPE_UIH;
+    frame_hdr->length   = (LENGTH_INDICATOR_OCTET | (size << 1));
+
+    memmove(&(frame_hdr->information[0]), buffer, size);
+
+    uint8_t* fcs_pos = (&(frame_hdr->information[0]) + size);
+    *fcs_pos         = fcs_calculate(&(Mux::_tx_context.buffer[FRAME_ADDRESS_FIELD_INDEX]), FCS_INPUT_LEN);
+    *(++fcs_pos)     = FLAG_SEQUENCE_OCTET;
+
+    _tx_context.bytes_remaining = UIH_FRAME_MIN_LEN + size;
+}
+
+
+void Mux::tx_noretransmit_entry_run()
+{
+    write_do();
+}
+
+
+ssize_t Mux::user_data_tx(uint8_t dlci_id, const void* buffer, size_t size)
+{
+    MBED_ASSERT(size <=
```

Comment:
You seem to be providing the basic "unstructured octect stream" convergance layer, which means you should be acting as a stream file handle. You must accept any size here.

If we accept the non-blocking limitation (which will work with PPP and the AT Parser), I believe that means you just need to limit the size to your maximum, and return how much you actually managed to fit in 1 frame. Any user that knows they're using a non-blocking FileHandle will cope with that.

For blocking (eg if you just want to printf from the C library over the mux), you would need to construct a loop to write however many frames are necessary.

I would say blocking really must be implemented if this is to be generally useful, but that can wait until later. You do need to make sure `poll()` works though. Is the lack of `poll` why the ATHandler is spinning?

Reply:
IOTCELL-566 created for possible future implementation change regarding this feedback.

Blocking mode is not supported by the current implementation and for my knowledge no plans to implement that in the future.

Regarding "ATHandler is spinning" feedback: Propably managed by IOTCELL-556: added comment there for tracking

Reply:
The poll() functionality and write size functionality are clearly broken and not acceptable for merging to master as-is. This mux has one job to do, and that's to present itself as a functioning virtual serial port. 

The limitation on no blocking support is something that could be documented, but the non-blocking support has to actually work correctly.

How does this even work with PPP as it is? I guess it doesn't?

Reply:
On future blocking support, I think it would make sense to introduce some sort of "make blocking from non-blocking" helper to avoid having to implement extra logic for blocking in every non-blocking-capable `FileHandle` provider.

That's the way sockets work, and it's super convenient - you can just keep returning -EAGAIN and not worry whether it's blocking mode or not. `FileHandle` just lacks that because it historically was always blocking-only and non-blocking was bolted on as an extension afterwards. 

I'll look into what this would look like. It would be continuing work from the "poll wake" stuff here: https://github.com/kjbracey-arm/mbed-os/commits/cv_cs

Reply:
We have created JIRA tasks for future improvements and no implementation changes are planned at this point.

Do you consider that these are blocking the PR merge at this point?

Reply:
Yes. See my comment below.

Reply:
Fixed with commits:
https://github.com/AriParkkila/mbedos/commit/1245c14de1dc8c7567553dde4ba3a30eb6e8495e
https://github.com/AriParkkila/mbed-os/commit/ee5d27f53603fec632c37fa7bd7e4a8aca28de81